File: src/main/java/Mp4DefaultTrack.java
import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public final class Mp4DefaultTrack {

    // MP4 box types as big-endian FourCC ints.
    static final int MOOV = ('m' << 24) | ('o' << 16) | ('o' << 8) | 'v';
    static final int TRAK = ('t' << 24) | ('r' << 16) | ('a' << 8) | 'k';
    static final int TKHD = ('t' << 24) | ('k' << 16) | ('h' << 8) | 'd';
    static final int MDIA = ('m' << 24) | ('d' << 16) | ('i' << 8) | 'a';
    static final int MDHD = ('m' << 24) | ('d' << 16) | ('h' << 8) | 'd';
    static final int HDLR = ('h' << 24) | ('d' << 16) | ('l' << 8) | 'r';
    static final int MINF = ('m' << 24) | ('i' << 16) | ('n' << 8) | 'f';
    static final int STBL = ('s' << 24) | ('t' << 16) | ('b' << 8) | 'l';
    static final int STSD = ('s' << 24) | ('t' << 16) | ('s' << 8) | 'd';
    // hdlr handler subtypes
    static final int VIDE = ('v' << 24) | ('i' << 16) | ('d' << 8) | 'e';
    static final int SOUN = ('s' << 24) | ('o' << 16) | ('u' << 8) | 'n';
    static final int SUBT = ('s' << 24) | ('u' << 16) | ('b' << 8) | 't';
    static final int SBTL = ('s' << 24) | ('b' << 16) | ('t' << 8) | 'l';
    static final int TEXT = ('t' << 24) | ('e' << 16) | ('x' << 8) | 't';

    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("Usage: java Mp4TrackFlagTool <list|set|unset> <file> [trackId] [default|forced]");
            return;
        }

        String cmd = args[0];
        File file = new File(args[1]);
        String flag = args.length >= 4 ? args[3] : "default";
        Integer trackId = (args.length >= 3 && isAllDigits(args[2])) ? Integer.valueOf(args[2]) : null;

        if ("list".equals(cmd)) {
            listTracks(file);
        } else if ("set".equals(cmd)) {
            if (trackId == null) {
                throw new IllegalArgumentException("Missing trackId");
            }
            setFlag(file, trackId, flag, true);
        } else if ("unset".equals(cmd)) {
            if (trackId == null) {
                throw new IllegalArgumentException("Missing trackId");
            }
            setFlag(file, trackId, flag, false);
        } else {
            throw new IllegalArgumentException("Unknown command: " + cmd);
        }
    }

    static boolean isAllDigits(String s) {
        if (s == null || s.isEmpty()) {
            return false;
        }
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c < '0' || c > '9') {
                return false;
            }
        }
        return true;
    }

    static class TrackInfo {
        long tkhdOffset;
        long stsdOffset;
        long mdhdOffset;
        int trackId;
        boolean defaultFlag;
        boolean forcedFlag;
        String type;
        String language;
    }

    static void listTracks(File f) throws IOException {
        List<TrackInfo> tracks = readTracks(f);
        StringBuilder sb = new StringBuilder(64 + tracks.size() * 80);
        sb.append("[\n");
        for (int i = 0; i < tracks.size(); i++) {
            TrackInfo t = tracks.get(i);
            sb.append("	{\"id\": ")
                    .append(t.trackId)
                    .append(", \"type\": \"")
                    .append(t.type)
                    .append("\", \"lang\": \"")
                    .append(t.language)
                    .append("\", \"default\": ")
                    .append(t.defaultFlag)
                    .append(", \"forced\": ")
                    .append(t.forcedFlag)
                    .append('}');
            if (i != tracks.size() - 1) {
                sb.append(',');
            }
            sb.append('\n');
        }
        sb.append("]\n");
        System.out.print(sb);
    }

    static void setFlag(File f, int id, String flag, boolean val) throws IOException {
        List<TrackInfo> tracks = readTracks(f);
        try (RandomAccessFile raf = new RandomAccessFile(f, "rw")) {
            for (TrackInfo t : tracks) {
                if (t.trackId != id) {
                    continue;
                }
                if ("default".equals(flag)) {
                    patchTkhdFlag(raf, t.tkhdOffset, val);
                } else if ("forced".equals(flag)) {
                    patchStsdFlag(raf, t.stsdOffset, val);
                } else {
                    throw new IllegalArgumentException("Unknown flag: " + flag);
                }
            }
        }
    }

    static long readUInt32(RandomAccessFile raf) throws IOException {
        return raf.readInt() & 0xffffffffL;
    }

    static List<TrackInfo> readTracks(File f) throws IOException {
        List<TrackInfo> tracks = new ArrayList<>();
        try (RandomAccessFile raf = new RandomAccessFile(f, "r")) {
            long fileLen = raf.length();
            while (raf.getFilePointer() + 8 <= fileLen) {
                long pos = raf.getFilePointer();
                long size = readUInt32(raf);
                int type = raf.readInt();
                if (size == 1) {
                    size = raf.readLong();
                } else if (size == 0) {
                    size = fileLen - pos;
                }
                if (type == MOOV) {
                    parseMoov(raf, pos, size, tracks);
                }
                if (size < 8) {
                    break;
                } else {
                    raf.seek(pos + size);
                }
            }
        }
        return tracks;
    }

    static void parseMoov(RandomAccessFile raf, long start, long size, List<TrackInfo> tracks) throws IOException {
        long end = start + size;
        raf.seek(start + 8);
        while (raf.getFilePointer() + 8 <= end) {
            long pos = raf.getFilePointer();
            long boxSize = readUInt32(raf);
            int type = raf.readInt();
            if (boxSize == 1) {
                boxSize = raf.readLong();
            } else if (boxSize == 0) {
                boxSize = end - pos;
            }
            if (type == TRAK) {
                TrackInfo info = parseTrak(raf, pos, boxSize);
                if (info != null) {
                    tracks.add(info);
                }
            }
            if (boxSize < 8) {
                break;
            } else {
                raf.seek(pos + boxSize);
            }
        }
    }

    static TrackInfo parseTrak(RandomAccessFile raf, long trakStart, long trakSize) throws IOException {
        TrackInfo info = new TrackInfo();
        long end = trakStart + trakSize;
        raf.seek(trakStart + 8);
        while (raf.getFilePointer() + 8 <= end) {
            long pos = raf.getFilePointer();
            long boxSize = readUInt32(raf);
            int type = raf.readInt();
            if (boxSize == 1) {
                boxSize = raf.readLong();
            } else if (boxSize == 0) {
                boxSize = end - pos;
            }
            if (type == TKHD) {
                info.tkhdOffset = pos + 8;
                raf.seek(info.tkhdOffset);
                int version = raf.readUnsignedByte();
                int flag1 = raf.readUnsignedByte();
                int flag2 = raf.readUnsignedByte();
                int flag3 = raf.readUnsignedByte();
                int flags = (flag1 << 16) | (flag2 << 8) | flag3;
                info.defaultFlag = (flags & 1) != 0;
                if (version == 1) {
                    raf.skipBytes(16);
                } else {
                    raf.skipBytes(8);
                }
                info.trackId = raf.readInt();
                raf.seek(pos + boxSize);
            } else if (type == MDIA) {
                parseMdia(raf, pos, boxSize, info);
            } else {
                raf.seek(pos + boxSize);
            }
        }
        if (info.trackId == 0) {
            return null;
        } else {
            return info;
        }
    }

    static String decodeMp4Language(int packed) {
        if (packed == 0) {
            return null;
        }
        char c1 = (char) (((packed >> 10) & 0x1F) + 0x60);
        char c2 = (char) (((packed >> 5) & 0x1F) + 0x60);
        char c3 = (char) ((packed & 0x1F) + 0x60);
        return "" + c1 + c2 + c3;
    }

    static void parseMdia(RandomAccessFile raf, long mdiaStart, long mdiaSize, TrackInfo info) throws IOException {
        byte[] header = new byte[8];
        long end = mdiaStart + mdiaSize;
        raf.seek(mdiaStart + 8);
        while (raf.getFilePointer() + 8 < end) {
            long pos = raf.getFilePointer();
            if (raf.read(header) != 8) {
                break;
            }
            int size = readInt(header, 0);
            int type = readInt(header, 4);
            if (size < 8) {
                break;
            }
            if (type == MDHD) {
                info.mdhdOffset = pos + 8;
                raf.seek(info.mdhdOffset);
                int version = raf.readUnsignedByte(); // 1 byte
                raf.skipBytes(3); // skip remaining flags (3 bytes)
                long creationTime = (version == 1) ? raf.readLong() : raf.readInt();
                long modificationTime = (version == 1) ? raf.readLong() : raf.readInt();
                int timescale = raf.readInt();
                long duration = (version == 1) ? raf.readLong() : raf.readInt();
                int langPacked = raf.readUnsignedShort(); // 2 bytes for ISO-639-2/T code
                info.language = decodeMp4Language(langPacked);
            } else if (type == HDLR) {
                raf.skipBytes(8); // version+flags+predefined
                byte[] htype = new byte[4];
                raf.readFully(htype);
                int subtype = readInt(htype, 0);
                if (subtype == VIDE) {
                    info.type = "video";
                } else if (subtype == SOUN) {
                    info.type = "audio";
                } else if (subtype == SUBT || subtype == SBTL || subtype == TEXT) {
                    info.type = "subtitle";
                } else {
                    info.type = new String(htype, StandardCharsets.US_ASCII);
                }
            } else if (type == MINF) {
                parseMinf(raf, pos, size, info);
            }
            raf.seek(pos + size);
        }
    }

    static void parseMinf(RandomAccessFile raf, long minfStart, int minfSize, TrackInfo info) throws IOException {
        byte[] header = new byte[8];
        long end = minfStart + minfSize;
        raf.seek(minfStart + 8);
        while (raf.getFilePointer() + 8 < end) {
            long pos = raf.getFilePointer();
            if (raf.read(header) != 8) {
                break;
            }
            int size = readInt(header, 0);
            int type = readInt(header, 4);
            if (size < 8) {
                break;
            }
            if (type == STBL) {
                parseStbl(raf, pos, size, info);
            }
            raf.seek(pos + size);
        }
    }

    static void parseStbl(RandomAccessFile raf, long stblStart, int stblSize, TrackInfo info) throws IOException {
        byte[] header = new byte[8];
        long end = stblStart + stblSize;
        raf.seek(stblStart + 8);
        while (raf.getFilePointer() + 8 < end) {
            long pos = raf.getFilePointer();
            if (raf.read(header) != 8) {
                break;
            }
            int size = readInt(header, 0);
            int type = readInt(header, 4);
            if (size < 8) {
                break;
            }
            if (type == STSD) {
                info.stsdOffset = pos + 8;
                raf.seek(info.stsdOffset + 8); // skip version+flags+entryCount
                byte[] entryHeader = new byte[8];
                if (raf.read(entryHeader) == 8) {
                    String sampleType = new String(entryHeader, 4, 4, StandardCharsets.US_ASCII);
                    info.forcedFlag = sampleType.toLowerCase(java.util.Locale.ROOT).contains("forced");
                }
            }
            raf.seek(pos + size);
        }
    }

    static void patchTkhdFlag(RandomAccessFile raf, long tkhdOffset, boolean set) throws IOException {
        raf.seek(tkhdOffset + 1);
        int f1 = raf.readUnsignedByte(), f2 = raf.readUnsignedByte(), f3 = raf.readUnsignedByte();
        int flags = (f1 << 16) | (f2 << 8) | f3;
        if (set) {
            flags |= 1;
        } else {
            flags &= ~1;
        }
        raf.seek(tkhdOffset + 1);
        raf.writeByte((flags >> 16) & 0xff);
        raf.writeByte((flags >> 8) & 0xff);
        raf.writeByte(flags & 0xff);
    }

    static void patchStsdFlag(RandomAccessFile raf, long stsdOffset, boolean set) throws IOException {
        raf.seek(stsdOffset + 8);
        byte[] entryHeader = new byte[8];
        raf.read(entryHeader);
        String type = new String(entryHeader, 4, 4, StandardCharsets.US_ASCII);
        if (set && !type.endsWith("fcd ")) {
            type = "fcd ";
        }
        raf.seek(stsdOffset + 12);
        raf.write(type.getBytes(StandardCharsets.US_ASCII), 0, 4);
    }

    static String decodeLang(int bits) {
        char a = (char) (((bits >> 10) & 31) + 0x60);
        char b = (char) (((bits >> 5) & 31) + 0x60);
        char c = (char) ((bits & 31) + 0x60);
        return "" + a + b + c;
    }

    static int readInt(byte[] buf, int off) {
        return ((buf[off] & 0xff) << 24) | ((buf[off + 1] & 0xff) << 16) | ((buf[off + 2] & 0xff) << 8)
                | (buf[off + 3] & 0xff);
    }
}

File: src/main/rust/Mp4DefaultTrack.rs
use std::env;
use std::fs::{File, OpenOptions};
use std::io::{self, Read, Seek, SeekFrom, Write};
use std::path::Path;
use std::convert::TryInto;

#[derive(Default, Debug)]
struct Track {
    tkhd_offset: Option<u64>,
    stsd_offset: Option<u64>,
    mdhd_offset: Option<u64>,
    track_id: u32,
    default_flag: bool,
    forced_flag: bool,
    typ: Option<String>,
    lang: Option<String>,
}

/// Reads the next MP4 box header: returns (box_size, type_bytes, header_size).
fn read_header(f: &mut File, max_size: u64) -> io::Result<Option<(u64, [u8; 4], u64)>> {
    let mut header = [0u8; 8];
    match f.read_exact(&mut header) {
        Ok(_) => {}
        Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(e) => return Err(e),
    }
    
    let mut box_size = u32::from_be_bytes(header[0..4].try_into().unwrap()) as u64;
    let typ: [u8; 4] = header[4..8].try_into().unwrap();
    let mut hdr_size = 8;

    if box_size == 1 {
        let mut ext = [0u8; 8];
        f.read_exact(&mut ext)?;
        box_size = u64::from_be_bytes(ext);
        hdr_size = 16;
    } else if box_size == 0 {
        box_size = max_size; // Extends to end of file/parent container
    }

    Ok(Some((box_size, typ, hdr_size)))
}

fn decode_mp4_language(packed: u16) -> Option<String> {
    if packed == 0 {
        return None;
    }
    let bytes = [
        ((packed >> 10) & 0x1F) as u8 + 0x60,
        ((packed >> 5) & 0x1F) as u8 + 0x60,
        (packed & 0x1F) as u8 + 0x60,
    ];
    String::from_utf8(bytes.to_vec()).ok()
}

fn parse_mp4(f: &mut File) -> io::Result<Vec<Track>> {
    let file_len = f.seek(SeekFrom::End(0))?;
    let mut tracks = Vec::new();
    let mut offset = 0;

    while offset + 8 <= file_len {
        f.seek(SeekFrom::Start(offset))?;
        let Some((box_size, typ, hdr_size)) = read_header(f, file_len - offset)? else { break };
        if box_size < hdr_size { break; }

        if typ == *b"moov" {
            parse_moov(f, offset + hdr_size, offset + box_size, &mut tracks)?;
        }
        offset += box_size;
    }

    Ok(tracks)
}

fn parse_moov(f: &mut File, mut offset: u64, limit: u64, tracks: &mut Vec<Track>) -> io::Result<()> {
    while offset + 8 <= limit {
        f.seek(SeekFrom::Start(offset))?;
        let Some((box_size, typ, hdr_size)) = read_header(f, limit - offset)? else { break };
        if box_size < hdr_size { break; }

        if typ == *b"trak" {
            if let Some(t) = parse_trak(f, offset + hdr_size, offset + box_size)? {
                tracks.push(t);
            }
        }
        offset += box_size;
    }
    Ok(())
}

fn parse_trak(f: &mut File, mut offset: u64, limit: u64) -> io::Result<Option<Track>> {
    let mut info = Track::default();
    
    while offset + 8 <= limit {
        f.seek(SeekFrom::Start(offset))?;
        let Some((box_size, typ, hdr_size)) = read_header(f, limit - offset)? else { break };
        if box_size < hdr_size { break; }

        if typ == *b"tkhd" {
            info.tkhd_offset = Some(offset + hdr_size);
            f.seek(SeekFrom::Start(offset + hdr_size))?;
            
            let mut tkhd_data = [0u8; 4];
            f.read_exact(&mut tkhd_data)?;
            let version = tkhd_data[0];
            let flags = ((tkhd_data[1] as u32) << 16) | ((tkhd_data[2] as u32) << 8) | (tkhd_data[3] as u32);
            info.default_flag = (flags & 1) != 0;

            if version == 1 {
                f.seek(SeekFrom::Current(16))?; // Skip creation/mod
            } else {
                f.seek(SeekFrom::Current(8))?;
            }
            
            let mut track_id_buf = [0u8; 4];
            f.read_exact(&mut track_id_buf)?;
            info.track_id = u32::from_be_bytes(track_id_buf);
            
        } else if typ == *b"mdia" {
            parse_mdia(f, offset + hdr_size, offset + box_size, &mut info)?;
        }
        
        offset += box_size;
    }

    if info.track_id != 0 { Ok(Some(info)) } else { Ok(None) }
}

fn parse_mdia(f: &mut File, mut offset: u64, limit: u64, info: &mut Track) -> io::Result<()> {
    while offset + 8 <= limit {
        f.seek(SeekFrom::Start(offset))?;
        let Some((box_size, typ, hdr_size)) = read_header(f, limit - offset)? else { break };
        if box_size < hdr_size { break; }

        if typ == *b"mdhd" {
            info.mdhd_offset = Some(offset + hdr_size);
            f.seek(SeekFrom::Start(offset + hdr_size))?;
            
            let mut ver = [0u8; 1];
            f.read_exact(&mut ver)?;
            f.seek(SeekFrom::Current(3))?; // Skip flags
            
            if ver[0] == 1 {
                f.seek(SeekFrom::Current(16 + 4 + 8))?; // ctime + mtime + timescale + duration
            } else {
                f.seek(SeekFrom::Current(8 + 4 + 4))?;
            }
            
            let mut lang_buf = [0u8; 2];
            f.read_exact(&mut lang_buf)?;
            info.lang = decode_mp4_language(u16::from_be_bytes(lang_buf));
            
        } else if typ == *b"hdlr" {
            f.seek(SeekFrom::Start(offset + hdr_size + 8))?; // pos + version+flags(4) + pre-defined(4)
            let mut hbuf = [0u8; 4];
            f.read_exact(&mut hbuf)?;
            info.typ = Some(match &hbuf {
                b"vide" => "video".to_string(),
                b"soun" => "audio".to_string(),
                b"subt" | b"sbtl" | b"text" => "subtitle".to_string(),
                _ => String::from_utf8_lossy(&hbuf).into_owned(),
            });
            
        } else if typ == *b"minf" {
            parse_minf(f, offset + hdr_size, offset + box_size, info)?;
        }
        offset += box_size;
    }
    Ok(())
}

fn parse_minf(f: &mut File, mut offset: u64, limit: u64, info: &mut Track) -> io::Result<()> {
    while offset + 8 <= limit {
        f.seek(SeekFrom::Start(offset))?;
        let Some((box_size, typ, hdr_size)) = read_header(f, limit - offset)? else { break };
        if box_size < hdr_size { break; }

        if typ == *b"stbl" {
            parse_stbl(f, offset + hdr_size, offset + box_size, info)?;
        }
        offset += box_size;
    }
    Ok(())
}

fn parse_stbl(f: &mut File, mut offset: u64, limit: u64, info: &mut Track) -> io::Result<()> {
    while offset + 8 <= limit {
        f.seek(SeekFrom::Start(offset))?;
        let Some((box_size, typ, hdr_size)) = read_header(f, limit - offset)? else { break };
        if box_size < hdr_size { break; }

        if typ == *b"stsd" {
            info.stsd_offset = Some(offset + hdr_size);
            f.seek(SeekFrom::Start(offset + hdr_size + 8))?; // Skip version+flags + entry count
            
            let mut entry_header = [0u8; 8];
            if f.read_exact(&mut entry_header).is_ok() {
                let sample_type = &entry_header[4..8];
                // Zero allocation forced-flag check
                info.forced_flag = sample_type[0..3].eq_ignore_ascii_case(b"fcd") 
                                || sample_type[1..4].eq_ignore_ascii_case(b"fcd");
            }
        }
        offset += box_size;
    }
    Ok(())
}

fn patch_tkhd_flag(f: &mut File, tkhd_offset: u64, set: bool) -> io::Result<()> {
    f.seek(SeekFrom::Start(tkhd_offset + 1))?;
    let mut b = [0u8; 3];
    f.read_exact(&mut b)?;
    let mut flags = ((b[0] as u32) << 16) | ((b[1] as u32) << 8) | (b[2] as u32);
    
    if set { flags |= 1; } else { flags &= !1; }
    
    let out = [((flags >> 16) & 0xff) as u8, ((flags >> 8) & 0xff) as u8, (flags & 0xff) as u8];
    f.seek(SeekFrom::Start(tkhd_offset + 1))?;
    f.write_all(&out)?;
    Ok(())
}

fn patch_stsd_forced(f: &mut File, stsd_offset: u64, set: bool) -> io::Result<()> {
    if set {
        let entry_header_pos = stsd_offset + 16;
        f.seek(SeekFrom::Start(entry_header_pos + 4))?;
        f.write_all(b"fcd ")?;
    }
    Ok(())
}

fn print_list(tracks: &[Track]) {
    println!("[");
    let len = tracks.len();
    for (i, t) in tracks.iter().enumerate() {
        let typ = t.typ.as_deref().unwrap_or("unknown");
        let lang = match &t.lang {
            Some(s) => format!("\"{}\"", s),
            None => "null".to_string(),
        };
        let comma = if i + 1 == len { "" } else { "," };
        println!("\t{{\"id\": {}, \"type\": \"{}\", \"lang\": {}, \"default\": {}, \"forced\": {}}}{}",
                 t.track_id, typ, lang, t.default_flag, t.forced_flag, comma);
    }
    println!("]");
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} <list|set|unset> <file> [trackId] [default|forced]", args[0]);
        std::process::exit(2);
    }

    let cmd = &args[1];
    let path = Path::new(&args[2]);
    let is_list = cmd == "list";
    let set_flag = cmd == "set";

    let mut file = match OpenOptions::new().read(true).write(!is_list).open(path) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("Error opening file: {}", e);
            std::process::exit(1);
        }
    };

    let tracks = match parse_mp4(&mut file) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("Error parsing file: {}", e);
            std::process::exit(1);
        }
    };

    if is_list {
        print_list(&tracks);
        std::process::exit(0);
    }

    if args.len() < 5 {
        eprintln!("Usage: {} set|unset <file> <trackId> <default|forced>", args[0]);
        std::process::exit(2);
    }

    let tid: u32 = match args[3].parse() {
        Ok(n) => n,
        Err(_) => {
            eprintln!("Invalid trackId: {}", args[3]);
            std::process::exit(2);
        }
    };
    let flag = &args[4];

    let mut found = false;
    for t in tracks {
        if t.track_id == tid {
            found = true;
            if flag == "default" {
                if let Some(off) = t.tkhd_offset {
                    if let Err(e) = patch_tkhd_flag(&mut file, off, set_flag) {
                        eprintln!("Error patching tkhd: {}", e);
                        std::process::exit(1);
                    }
                }
            } else if flag == "forced" {
                if let Some(off) = t.stsd_offset {
                    if let Err(e) = patch_stsd_forced(&mut file, off, set_flag) {
                        eprintln!("Error patching stsd: {}", e);
                        std::process::exit(1);
                    }
                }
            } else {
                eprintln!("Unknown flag: {}", flag);
                std::process::exit(2);
            }
            break;
        }
    }

    if !found {
        eprintln!("Track {} not found", tid);
        std::process::exit(1);
    }
}
